sqlx-core/src/pool/inner.rs

```diff
@@ -169,7 +170,19 @@ impl<DB: Database> SharedPool<DB> {
             self.options.connect_timeout,
             async {
                 loop {
-                    let permit = self.semaphore.acquire(1).await;
+                    // Decorate the semaphore permit acquisition to record the
+                    // duration it takes to be granted (or timed out).
+                    let fut = self.semaphore.acquire(1);
+                    let permit = if let Some(obs) = &self.options.metric_observer {
+                        // A metric observer is registered, so instrument the
+                        // semaphore wait.
+                        PollWallClockRecorder::new(
+                            |d| obs.permit_wait_time(d), 
+                            fut,
+                        ).await
+                    } else {
+                        fut.await
+                    };
```

Comment:
This seems a bit clunky.

We could implement a `NopObserver` and do away with the `Option`, but because this is a `dyn` I suspect the NOP wouldn't be free.

Reply:
It looks like in specific cases, LLVM can inline calls through `dyn`: https://play.rust-lang.org/?version=nightly&mode=release&edition=2021&gist=5c10929c5507f03b98638f457870521d

I think this is fine though, as it saves an `Arc` allocation if the feature is not used.

Reply:
You could also just implement `PoolMetricsObserver` for `Option<T> where T: PoolMetricsObserver` and move the complexity there.

Reply:
The `Option` impl idea is spot on - I'll do this shortly.

Reply:
Done in 1d25e95a